benchmark/spmv/cuda_linops.hpp

```diff
@@ -0,0 +1,463 @@
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef __CUDA_LINOPS__
+#define __CUDA_LINOPS__
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <cuda_runtime.h>
+#include <cusparse.h>
+#include <memory>
+
+
+#include "cuda/base/cusparse_bindings.hpp"
+#include "cuda/base/device_guard.hpp"
+
+
+namespace detail {
+
+
+class CuspBase : public gko::LinOp {
+public:
+    void apply_impl(const gko::LinOp *, const gko::LinOp *, const gko::LinOp *,
+                    gko::LinOp *) const override
+    {}
+
+    cusparseMatDescr_t get_descr() const { return this->descr_.get(); }
+
+    const gko::CudaExecutor *get_gpu_exec() const { return gpu_exec; }
+
+protected:
+    CuspBase(std::shared_ptr<const gko::Executor> exec,
+             const gko::dim<2> &size = gko::dim<2>{})
+        : gko::LinOp(exec, size)
+    {
+        gpu_exec = dynamic_cast<const gko::CudaExecutor *>(exec.get());
+        if (gpu_exec == nullptr) {
+            GKO_NOT_IMPLEMENTED;
+        }
+        this->initialize_descr();
+    }
+
+    CuspBase &operator=(const CuspBase &other)
+    {
+        this->gpu_exec = other.get_gpu_exec();
+        this->initialize_descr();
+        return *this;
+    }
+
+    void initialize_descr()
+    {
+        const auto id = this->gpu_exec->get_device_id();
+        gko::device_guard g{id};
+        this->descr_ = handle_manager<cusparseMatDescr>(
+            gko::kernels::cuda::cusparse::create_mat_descr(),
+            [id](cusparseMatDescr_t descr) {
+                gko::device_guard g{id};
+                gko::kernels::cuda::cusparse::destroy(descr);
+            });
+    }
+
+private:
+    const gko::CudaExecutor *gpu_exec;
+    template <typename T>
+    using handle_manager = std::unique_ptr<T, std::function<void(T *)>>;
+    handle_manager<cusparseMatDescr> descr_;
+};
+
+
+template <typename ValueType = gko::default_precision,
+          typename IndexType = gko::int32>
+class CuspCsrmp
+    : public gko::EnableLinOp<CuspCsrmp<ValueType, IndexType>, CuspBase>,
+      public gko::ReadableFromMatrixData<ValueType, IndexType>,
+      public gko::EnableCreateMethod<CuspCsrmp<ValueType, IndexType>> {
+    friend class gko::EnableCreateMethod<CuspCsrmp>;
+
+public:
+    using csr = gko::matrix::Csr<ValueType, IndexType>;
+    using mat_data = gko::matrix_data<ValueType, IndexType>;
+
+    void read(const mat_data &data) override
+    {
+        csr_->read(data);
+        this->set_size(gko::dim<2>{csr_->get_size()});
+    }
+
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override
+    {
+        auto dense_b = gko::as<gko::matrix::Dense<double>>(b);
+        auto dense_x = gko::as<gko::matrix::Dense<ValueType>>(x);
+        auto db = dense_b->get_const_values();
+        auto dx = dense_x->get_values();
+        auto alpha = gko::one<ValueType>();
+        auto beta = gko::zero<ValueType>();
```

Comment:
alpha and beta are on host memory, so it returns the error
`Unrecoverable CUDA error on device 0 in raw_free: cudaErrorIllegalAddress: an illegal memory access was encountered`

Reply:
So it needs to put them into device memory

Reply:
Ah, I did not see they were passed by pointers. That makes sense. I was also running the test on a V100 and did not get these errors due to UVM. This is now fixed, thanks !